integration/docker/csi/alluxio/nodeserver.go

```diff
@@ -87,45 +103,203 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+ */
+func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpublishVolumeRequest) (*csi.NodeUnpublishVolumeResponse, error) {
+	targetPath := req.GetTargetPath()
+	command := exec.Command("umount", targetPath)
+	glog.V(4).Infoln(command)
 	stdoutStderr, err := command.CombinedOutput()
 	if err != nil {
 		glog.V(3).Infoln(err)
 	}
 	glog.V(4).Infoln(string(stdoutStderr))
 
-	err = mount.CleanupMountPoint(req.GetTargetPath(), mount.New(""), false)
+	err = mount.CleanupMountPoint(targetPath, mount.New(""), false)
 	if err != nil {
 		glog.V(3).Infoln(err)
+	} else {
+		glog.V(4).Infof("Succeed in unmounting %s", targetPath)
 	}
-
 	return &csi.NodeUnpublishVolumeResponse{}, nil
 }
 
-func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
-	return &csi.NodeUnstageVolumeResponse{}, nil
-}
-
+/*
+ * If mount in node server, do nothing
+ * If mount in separate pod, mount Alluxio to global mount point
+ */
 func (ns *nodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	if req.GetVolumeContext()["mountInPod"] == "true" {
+		ns.mutex.Lock()
+		defer ns.mutex.Unlock()
+
+		glog.V(4).Infoln("Creating Alluxio-fuse pod and mounting Alluxio to global mount point.")
+		fusePod, err := getAndCompleteFusePodObj(ns.nodeId, req)
+		if err != nil {
+			return nil, err
+		}
+		if _, err := ns.client.CoreV1().Pods("default").Create(fusePod); err != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to launch Fuse Pod at %v.\n%v", ns.nodeId, err.Error())
+		}
+		glog.V(4).Infoln("Successfully creating Fuse pod.")
+	}
 	return &csi.NodeStageVolumeResponse{}, nil
 }
 
+func getAndCompleteFusePodObj(nodeId string, req *csi.NodeStageVolumeRequest) (*v1.Pod, error) {
+	csiFuseYaml, err := ioutil.ReadFile("/opt/alluxio/integration/kubernetes/csi/alluxio-csi-fuse.yaml")
+	if err != nil {
+		glog.V(4).Info("csi-fuse config yaml file not found")
+		return nil, status.Errorf(codes.NotFound, "csi-fuse config yaml file not found: %v", err.Error())
+	}
+	csiFuseObj, grpVerKind, err := scheme.Codecs.UniversalDeserializer().Decode(csiFuseYaml, nil, nil)
+	if err != nil {
+		glog.V(4).Info("Failed to decode csi-fuse config yaml file")
+		return nil, status.Errorf(codes.Internal, "Failed to decode csi-fuse config yaml file.\n", err.Error())
+	}
+	// Only support Fuse Pod for now
+	if grpVerKind.Kind != "Pod" {
+		glog.V(4).Info("csi-fuse only support pod. %v found.")
+		return nil, status.Errorf(codes.InvalidArgument, "csi-fuse only support Pod. %v found.\n%v", grpVerKind.Kind, err.Error())
+	}
+	csiFusePodObj := csiFuseObj.(*v1.Pod)
+
+	// Add nodeName to pod name for uniqueness
+	if nodeId == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "nodeID is missing in the csi setup.\n%v", err.Error())
+	}
+	csiFusePodObj.Name = csiFusePodObj.Name + nodeId
```

Comment:
What if multi-pod mount the different volume in the same node? Will they use the same pod name?

Reply:
Good point. Thank you.